clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,83 @@
+use crate::utils::{is_expn_of, match_qpath, span_note_and_lint};
+use rustc::hir::{Expr, ExprKind, Mutability, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function calls that mutate an object in `debug_assert!`, `debug_assert_eq!` and
+    /// `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds the code won't be executed and therefore can alter
```

Comment:
"alter the execution flow" doens't sound nice, does it? Not sure what to say else here

Reply:
"In release builds `debug_assert` macros are optimized out by the compiler. Therefore mutating something in a `debug_assert` macro results in different behavior of the release and the debug build."